fix: add option for reusing a pending SSR render for subsequent similar requests (instead of fallback to CSR)

Platonn opened the pull request:
Before this PR:
When the html for a request was being rendered and in the meantime new requests arrived for the same rendering key (e.g. for the same URL), they were resolved immediately with a CSR fallback.

This PR preserves the previous default behavior, but also adds a new opt-in config that helps avoiding unnecessary CSR fallbacks.

The new opt-in config `reuseCurrentRendering` allows the subsequent requests to wait for the result of the pending render that was triggered by the 1st request. Finally, the all the waiting requests for the same rendering key (e.g. URL) will reuse the result of a single completed render.

Note 1: Same as previously, each request can time out independently (according to the `timeout` config) and then this request is resolved with a CSR fallback. However, with the new config `reuseCurrentRendering` enabled, other requests (which didn't time out yet) can still wait for the result of the render.

Note 2: The existing config `concurency` limits the number of the pending renders. So when config `reuseCurrentRendering` is enabled,  requests for the same rendering key take up just one concurrency slot, because they share just one pending render.

Note 3: Previously the `RenderinStrategy.ALWAYS_SSR` caused triggering of a new render for each request, even if there was a pending render for the same rendering key. This PR preserves the previous default behavior. However, with the new config `reuseCurrentRendering` enabled, the subsequent requests for the same key won't trigger new renders, but will wait for the pending render to complete.

closes #13623

znikola committed:
queue processing done

znikola committed:
Merge remote-tracking branch 'origin/develop' into feature/GH-13623

znikola committed:
updated log message

znikola committed:
backwards compat

znikola committed:
added the option

znikola committed:
added tests; added crawler test

znikola committed:
enable option by default

znikola committed:
Merge branch 'develop' into feature/GH-13623

znikola committed:
avoiding confussion aroung the units used for maxRenderTime

znikola committed:
added tests

znikola committed:
renamed queueprocessing indicator

znikola committed:
concurrency progress

znikola committed:
concurrency done

znikola committed:
renamed the option; renamed and consolidated tests

znikola committed:
Merge remote-tracking branch 'origin/develop' into feature/GH-13623

znikola committed:
cr: renamed private var

znikola committed:
cr: renamed infinite loop var

znikola committed:
typesafety in server.ts

znikola committed:
added a jsdoc

znikola committed:
Merge branch 'develop' into feature/GH-13623

znikola committed:
Merge branch 'develop' into feature/GH-13623

znikola committed:
Merge branch 'develop' into feature/GH-13623

znikola committed:
Merge branch 'develop' into feature/GH-13623

znikola committed:
cr

znikola committed:
Merge branch 'feature/GH-13623' of github.com:SAP/spartacus into feature/GH-13623

znikola committed:
Merge branch 'develop' into feature/GH-13623

znikola committed:
added queue count to the log message

znikola committed:
not exposing the internal ssrcallback type

znikola committed:
Merge branch 'develop' into feature/GH-13623

znikola committed:
reverted type changes; fixed a bug;

znikola committed:
change the comment style

Platonn committed:
attempt to simplify the conditions for ++ and -- of the current concurrency

Platonn committed:
flatten ifs, group handing current concurrency and init/clear waitingRenderCallbacks under one condition of isFirstRequestForKey

Platonn committed:
clear the waitingRenderCallbacks array in case of the first request passing maxRenderTimeout

Platonn committed:
deduce a request is the first for the key, based on the

Platonn committed:
fix: swap this.expressEngine and waitingRenderCallbacks.push based on the isFirstRequestForKey

Platonn committed:
assuming reuseCurrentRendering checks are unnecessary, now removing them

Platonn committed:
formatting and commenting

Platonn committed:
formatting and commenting 2

Platonn committed:
assuming the check for reuseCurrentRendering is not needed in isConcurrencyLimitExceeded, now removing it

Platonn committed:
fix: drop the remove 'request' param from isConcurrencyLimitExceeded

Platonn committed:
commenting

Platonn committed:
commenting

Platonn committed:
fix: remove unnecessary check for reuseCurrentRendering on invoking this.expressEnging

Platonn committed:
commenting

Platonn committed:
use Map instead of Record for waitingRenderCallbacks to avoid unnecessary dead entries with 'null'

Platonn committed:
commenting

Platonn committed:
narrow typing of waitingRenderCallbacks

Platonn committed:
add missing checks for the feature flag and extract concurrency handling logic to separate methods

Platonn committed:
idon't invoke the subsequent callbacks by the first callback, but instad create one wrapper callback to invoke all other (including the first request)

Platonn committed:
update the comment for the simplified semantics of the waitingRenderCallbacks

Platonn committed:
add unit tests

Platonn committed:
fix names of tests

Platonn committed:
commenting

Platonn committed:
add unit tests for custom renderingStrategyResolver function

Platonn committed:
test naming

Platonn committed:
New unit tests from branch feature/GH-13702 (Merge branch 'feature/GH-13702' of github.com:SAP/spartacus into feature/GH-13623--v2)

Platonn committed:
Merge branch 'feature/GH-13623--v2' of github.com:SAP/spartacus into feature/GH-13623--v2

Platonn committed:
added checks for the  state of currentConcurrency

Platonn committed:
Add a few new expects in existing tests (Merge branch 'feature/GH-13702' of github.com:SAP/spartacus into feature/GH-13623--v2)

Platonn committed:
update unit test for Processing many requests with the same result

Platonn committed:
add checks for currentConcurrency in existing tests

Platonn committed:
for readability of local tests, place timeout and renderTime in specific tests instead of shared consts

Platonn committed:
test: add checks for waitingRenderCallbacks and introduce helper test functions to assert waitingRenderCallbacks and currentConcurrency

Platonn committed:
test: should queue reuqests for the same rendering key even when the rendering strategy is ALWAYS_SSR

Platonn committed:
rename waitingRenderCallbacks to renderCallbacks + add comments

Platonn committed:
remove this.log('Processing N waiting SSR requests for URL')

Platonn committed:
test: add test for rendering cache - not to drop oldest element, when setting a new value for existing key

Platonn committed:
document that  will take up only one concurrency slot for one rendering key

Platonn commented on projects/storefrontapp/server.ts:
```diff
@@ -16,13 +17,14 @@ const express = require('express');
 const ssrOptions: SsrOptimizationOptions = {
   concurrency: 20,
   timeout: Number(process.env.SSR_TIMEOUT ?? 3000),
+  reuseCurrentRendering: true,
```
To be removed?

znikola replied:
I'd rather have it on by default in our own setup

znikola commented on core-libs/setup/ssr/optimized-engine/optimized-ssr-engine.ts:
```diff
@@ -214,4 +311,90 @@ export class OptimizedSsrEngine {
+    isFirstRequestForKey,
+  }: {
+    isFirstRequestForKey: boolean;
+  }): void {
```
Since the method is private, we can evolve it freely and have the param as an object (if there's a need). For now, for readability and simplicity sake, I suggest to just pass one param `isFirstRequestForKey: boolean`

znikola commented on core-libs/setup/ssr/optimized-engine/optimized-ssr-engine.ts:
```diff
@@ -214,4 +311,90 @@ export class OptimizedSsrEngine {
+  private decrementCurrentConcurrency({
+    isFirstRequestForKey,
+  }: {
+    isFirstRequestForKey: boolean;
```
The same as above

znikola commented on core-libs/setup/ssr/optimized-engine/optimized-ssr-engine.ts:
```diff
@@ -110,92 +174,125 @@ export class OptimizedSsrEngine {
     const request: Request = options.req;
     const response: Response = options.res || options.req.res;
 
+    if (this.returnCachedRender(request, callback)) {
+      this.log(`Render from cache (${request?.originalUrl})`);
+      return;
+    }
+    if (!this.shouldRender(request)) {
+      this.fallbackToCsr(response, filePath, callback);
+      return;
+    }
```
Nicely flattened 👍🏻

znikola commented on core-libs/setup/ssr/optimized-engine/optimized-ssr-engine.ts:
```diff
@@ -110,92 +174,125 @@ export class OptimizedSsrEngine {
+      );
+
+      if (this.ssrOptions?.reuseCurrentRendering) {
+        this.renderCallbacks.delete(renderingKey);
```
Nice. Are we covering this in a test?

Platonn replied:
Yes. Here https://github.com/SAP/spartacus/blob/feature/GH-13623--v2/core-libs/setup/ssr/optimized-engine/optimized-ssr-engine.spec.ts#L925-L937

znikola commented on core-libs/setup/ssr/optimized-engine/optimized-ssr-engine.spec.ts:
```diff
@@ -47,29 +48,40 @@ class TestEngineRunner {
 
+const getCurrentConcurrency = (
+  engineRunner: TestEngineRunner
+): { currentConcurrency: number } => {
```
Do we plan to return more parameters as part of the return type?

Platonn replied:
Due to a lot of assertions on numbers 0,1,2,3 inside one it() and because we check in turns: currentConcurrency as well as renderCallbacksCount, I deliberately wrapped the expected value in a meaningful object. Thanks to that the failed test doesn't tell "expected 3 to be 0", but expected { currentConcurrency: 3} to equal { currentConcurrency: 0 }

I suspect I'm solving the readability issue in a fancy way, but the real problem might be somewhere else (e.g. I perform too many `expect`s in one `it`)...

Platonn committed:
Merge branch 'develop' of github.com:SAP/spartacus into feature/GH-13623--v2

znikola commented on core-libs/setup/ssr/optimized-engine/optimized-ssr-engine.ts:
```diff
@@ -110,92 +174,125 @@ export class OptimizedSsrEngine {
+      this.fallbackToCsr(response, filePath, callback);
     }
+
+    // start rendering
```
This comment is a bit misleading because we don't actually start the rendering here, and also because of line 286 where we call the `startRendering()`. I suggest to remove it.

Platonn replied:
Will do

znikola commented on core-libs/setup/ssr/optimized-engine/optimized-ssr-engine.ts:
```diff
@@ -214,4 +311,90 @@ export class OptimizedSsrEngine {
+   *
+   * In case when the config `reuseCurrentRendering` is enabled and **if there is already a pending
+   * render task for the same rendering key**, it doesn't delegate a new render to Angular Universal.
+   * Instead, it waits for the current rendering to complete and then reuse the result for all waiting requests.
```
For clarity, I'd also mention here that it waits until each request's timeout expires.
Maybe something like:

```text
Instead, it waits for the current render to complete (respecting the timer option) and then reuse the result for all waiting requests.
```

Platonn replied:
Hmmm... To be precise, We don't invoke the render callbacks when the timeout passes. We wait until the render finishes and only then we invoke the callbacks. 
And in case we perceive the render as hanging (due to maxRenderTime), we forget those callbacks and never invoke them.

In ideal design, we should remove those callbacks from the collection when they timeout. Now we have to invoke the callback anyway when the render finishes (even after timeout) - because the callback is responsible for storing the result in the cache if the request timed out.

Further refactor seems quite complex, so we should plan it carefully. TBD

znikola commented on core-libs/setup/ssr/optimized-engine/optimized-ssr-engine.ts:
```diff
@@ -214,4 +311,90 @@ export class OptimizedSsrEngine {
+    renderCallback: SsrCallbackFn;
+    request: Request;
+    isFirstRequestForKey: boolean;
+  }) {
```
let's explicitly specify the function's return type to `void`

Platonn replied:
Will do

znikola commented on core-libs/setup/ssr/optimized-engine/optimized-ssr-engine.ts:
```diff
@@ -214,4 +311,90 @@ export class OptimizedSsrEngine {
+   * render task for the same rendering key**, it doesn't delegate a new render to Angular Universal.
+   * Instead, it waits for the current rendering to complete and then reuse the result for all waiting requests.
+   */
+  private startRender({
```
The name is at the same the perfect and confusing:
- perfect because it's short and describes exactly what the method does
- but it's confusing because it starts the rendering only for the first request for key (if the option is enabled ofc), but _not_ for the subsequent requests. I know this is already described in method's jsdoc, but it's still confusing unless you have a bigger picture. Something to consider and discuss. Maybe `manageRendering` could be a candidate.

znikola commented on core-libs/setup/ssr/optimized-engine/optimized-ssr-engine.ts:
```diff
@@ -214,4 +311,90 @@ export class OptimizedSsrEngine {
+      this.expressEngine(filePath, options, (err, html) => {
+        // Share the result of the render with all awaiting requests for the same key:
+
+        // Note: we access the Map at the moment of the render finished (don't store value it in a local variable),
```
Is there a typo in this comment? In this part: `don't store value it in a local variable`?

Platonn replied:
Thanks. should be: _don't store value ~it~ in a local variable_

Platonn replied:
Thanks, should be
```suggestion
        // Note: we access the Map at the moment of the render finished (don't store value in a local variable),
```

znikola commented on core-libs/setup/ssr/optimized-engine/optimized-ssr-engine.spec.ts:
```diff
@@ -586,4 +619,375 @@ describe('OptimizedSsrEngine', () => {
+
+    describe('when enabled', () => {
+      describe('multiple subsequent requests for the same rendering key should reuse the current render', () => {
+        it('and the first request should timeout', fakeAsync(() => {
```
The name of the test is not clear.
Are we testing the case when the first request (the one that started the render) times out?

znikola commented on core-libs/setup/ssr/optimized-engine/optimized-ssr-engine.spec.ts:
```diff
@@ -586,4 +619,375 @@ describe('OptimizedSsrEngine', () => {
+          // eventually the render succeeds and 3 remaining requests get the same response:
+          expect(engineRunner.renderCount).toEqual(1);
+          expect(engineRunner.renders).toEqual([
+            '',
```
Fallback comment? :)

znikola commented on core-libs/setup/ssr/optimized-engine/optimized-ssr-engine.spec.ts:
```diff
@@ -586,4 +619,375 @@ describe('OptimizedSsrEngine', () => {
+
+          expect(renderExceedMessageCount).toBe(2);
+          expect(engineRunner.renderCount).toEqual(0);
+          expect(engineRunner.renders).toEqual(['', '']);
```
Fallback comment? :)

znikola commented on core-libs/setup/ssr/optimized-engine/optimized-ssr-engine.spec.ts:
```diff
@@ -586,4 +619,375 @@ describe('OptimizedSsrEngine', () => {
+
+          tick(100);
+          expect(engineRunner.renderCount).toEqual(1);
+          expect(engineRunner.renders).toEqual(['', `${requestUrl}-0`]);
```
Fallback comment? :)

znikola commented on core-libs/setup/ssr/optimized-engine/optimized-ssr-engine.spec.ts:
```diff
@@ -586,4 +619,375 @@ describe('OptimizedSsrEngine', () => {
+          `SSR rendering exceeded timeout ${timeout}, fallbacking to CSR for ${requestUrl}`,
+          false
+        );
+        expect(engineRunner.renders).toEqual(['', '']);
```
Fallback comment? :)

znikola commented on core-libs/setup/ssr/optimized-engine/optimized-ssr-engine.spec.ts:
```diff
@@ -586,4 +619,375 @@ describe('OptimizedSsrEngine', () => {
+        );
+
+        expect(engineRunner.renderCount).toEqual(1);
+        expect(engineRunner.renders).toEqual(['', '']);
```
Fallback comment? :)

znikola commented on core-libs/setup/ssr/optimized-engine/optimized-ssr-engine.spec.ts:
```diff
@@ -586,4 +619,375 @@ describe('OptimizedSsrEngine', () => {
+        }));
+      });
+
+      it('should NOT queue the subsequent requests for a different rendering key', fakeAsync(() => {
```
Let's not use `queue` anymore, as we're using the map internally.

znikola committed:
Merge branch 'develop' into feature/GH-13623--v2

Platonn committed:
fixes after review

Platonn committed:
shorten one test

Platonn committed:
Merge branch 'feature/GH-13623--v2' of github.com:SAP/spartacus into feature/GH-13623--v2

znikola commented on core-libs/setup/ssr/optimized-engine/optimized-ssr-engine.ts:
```diff
@@ -8,6 +8,17 @@ import {
   SsrOptimizationOptions,
 } from './ssr-optimization-options';
 
+export type SsrCallbackFn = (
```
Adding this comment to a random place - as agreed, we will re-add a log statement.

Platonn committed:
fixes after review

Platonn committed:
fix: set maxRenderTimeout only once per render, but not once per every request (#13785)

- MAIN CHANGE: set maxRenderTimeout only once per render, but not once per every request
- change the log message to be more precise
   ```diff
   - N requests waiting for the render to complete (URL)
   + Request is waiting for the render to complete (URL)
   ```
   Note: some requests might be no longer waiting, because after timeout they were served with a CSR fallback
- rename `waitingForRender` variable to `requestTimeout`, becasue it was confusing - the timeout is related to the request, but not to the render
- by the way:
  - encapsulated all logic related to the render itself (execute the render, set max render time, handle concurrency of renders) inside one method `startRender`
  - get rid of the conditional logic for changing the `currentConcurrency` state
  - get rid of the variable/param `isFirstRequestForKey` (now we check whether someting is rendering in just one place) 
  - fix and unit-test: don't exceed the concurrency limit if the request is for a key that is alreaby being rendered
  - drop method `isRendering(request)`, as its parameter was confusing (we don't check if the reuqest is being rendered, but if there is a pending render for the renderingKey).

znikola committed:
Merge branch 'develop' into feature/GH-13623--v2

Platonn committed:
Merge branch 'develop' into feature/GH-13623--v2
